### ISSUE TITLE
How to make GpuInit() execute correctly in microsoft-standard-WSL2 environment?

### ISSUE THREAD

**Issue by zxl777** [alphacep/vosk-api#922]
```
uname -a
Linux DESKTOP-S7O3QUS 5.10.102.1-microsoft-standard-WSL2 #1 SMP Wed Mar 2 00:30:59 UTC 2022 x86_64 GNU/Linux
```

```
/usr/local/cuda/bin/nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_Mar__8_18:18:20_PST_2022
Cuda compilation tools, release 11.6, V11.6.124
Build cuda_11.6.r11.6/compiler.31057947_0
```

```
/usr/lib/wsl/lib/nvidia-smi
Thu Apr 14 14:12:46 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.60.02    Driver Version: 512.15       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0  On |                  N/A |
| 23%   36C    P0    49W / 307W |    443MiB /  8192MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from vosk import BatchModel, BatchRecognizer, GpuInit
>>> GpuInit()
>>> model = BatchModel()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/vosk/__init__.py", line 117, in __init__
    raise Exception("Failed to create a model")
Exception: Failed to create a model
>>> 
```

**Comment by nshmyrev** [comment]
I don't think we support BatchRecognizer on Windows yet. Only Linux.

**Comment by zxl777** [comment]
It's Linux, not Windows.

WSL2  = Windows Subsystem for Linux, 
Now WSL2 is basically the same as ubuntu and can also run docker.
I loaded the image of https://github.com/sskorol/vosk-api-gpu on WSL2, but it said that the GPU could not be found.

I was successful in running nvidia's docker container on this platform.
Refer to this URL: https://docs.nvidia.com/cuda/wsl-user-guide/index.html
This shows that the GPU driver and the docker environment can work.

I hope you can help with this so that a Windows PC with a GPU loaded can run the VOSK server in the background Linux.

Thanks.

**Comment by zxl777** [comment]
I run the container with https://github.com/sskorol/vosk-api-gpu and the error reported is that no GPU was found, but the GPU drivers are all installed correctly.
What parts do I need to check if I want to fix this?

**Comment by scotgopal** [comment]
```
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from vosk import BatchModel, BatchRecognizer, GpuInit
>>> GpuInit()
>>> model = BatchModel()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/dist-packages/vosk/__init__.py", line 117, in __init__
    raise Exception("Failed to create a model")
Exception: Failed to create a model
>>> 
```

Hey @zxl777 , were you able to overcome this stage and create a `BatchModel` object?

**Comment by sskorol** [comment]
@scotgopal see WSL instructions [here](https://github.com/sskorol/vosk-api-gpu#windows-11-with-wsl2). And a related discussion in this [issue](https://github.com/sskorol/vosk-api-gpu/issues/11).

**Comment by zxl777** [comment]
`docker run --rm -it --gpus=all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark`
Make sure that this command is executed successfully, it means that the environment has been set up correctly.

**Comment by scotgopal** [comment]
> @scotgopal see WSL instructions [here](https://github.com/sskorol/vosk-api-gpu#windows-11-with-wsl2). And a related discussion in this [issue](https://github.com/sskorol/vosk-api-gpu/issues/11).

Thanks @sskorol . I'm trying to run it locally without using docker.

> docker run --rm -it --gpus=all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark

@zxl777 I'm able to run this fine and got the following output.
```shell
$ docker run --rm -it --gpus=all nvcr.io/nvidia/k8s/cuda-sample:nbody nbody -gpu -benchmark
Run "nbody -benchmark [-numbodies=<numBodies>]" to measure performance.
        -fullscreen       (run n-body simulation in fullscreen mode)
        -fp64             (use double precision floating point values for simulation)
        -hostmem          (stores simulation data in host memory)
        -benchmark        (run benchmark to measure performance)
        -numbodies=<N>    (number of bodies (>= 1) to run in simulation)
        -device=<d>       (where d=0,1,2.... for the CUDA device to use)
        -numdevices=<i>   (where i=(number of CUDA devices > 0) to use for simulation)
        -compare          (compares simulation results running once on the default GPU and once on the CPU)
        -cpu              (run n-body simulation on the CPU)
        -tipsy=<file.bin> (load a tipsy model file for simulation)

NOTE: The CUDA Samples are not meant for performance measurements. Results may vary when GPU Boost is enabled.

> Windowed mode
> Simulation data stored in video memory
> Single precision floating point simulation
> 1 Devices used for simulation
GPU Device 0: "Pascal" with compute capability 6.1

> Compute 6.1 CUDA device: [NVIDIA GeForce GTX 1050]
6144 bodies, total time for 10 iterations: 5.457 ms
= 69.176 billion interactions per second
= 1383.524 single-precision GFLOP/s at 20 flops per interaction
```
Here are more helpful details on Vosk pkg version, model dir and error stacktrace:-
```shell
$ pip list | grep vosk
vosk                          0.3.38

$ ls .
model              test.wav              test_gpu_batch.py   test_simple.py   test_webvtt.py
model-en-us        test_alternatives.py  test_microphone.py  test_speaker.py  test_words.py
model-small-en-us  test_empty.py         test_nlsml.py       test_srt.py      vosk-model-en-us-0.22-lgraph.zip
output.wav         test_ffmpeg.py        test_reset.py       test_text.py     vosk-model-small-en-us-0.15.zip

$ tree model
model
├── README
├── am
│   ├── final.mdl
│   └── tree
├── conf
│   ├── mfcc.conf
│   └── model.conf
├── graph
│   ├── Gr.fst
│   ├── HCLr.fst
│   ├── disambig_tid.int
│   ├── phones
│   │   └── word_boundary.int
│   ├── phones.txt
│   └── words.txt
└── ivector
    ├── final.dubm
    ├── final.ie
    ├── final.mat
    ├── global_cmvn.stats
    ├── online_cmvn.conf
    └── splice.conf

$ python3
Python 3.8.10 (default, Mar 15 2022, 12:22:08)
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from vosk import GpuInit, BatchModel
>>> GpuInit()
>>> model = BatchModel()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/scotlinux/neura-gitlab/vosk-testing/venv/lib/python3.8/site-packages/vosk/__init__.py", line 117, in __init__
    raise Exception("Failed to create a model")
Exception: Failed to create a model
```

**Comment by sskorol** [comment]
@scotgopal did you build Vosk from sources with GPU support? Or you just installed an official package from PyPi? If latter, I don't believe GPU-version is published. You have to build it or just use docker.

**Comment by scotgopal** [comment]
> @scotgopal did you build Vosk from sources with GPU support? Or you just installed an official package from PyPi? If latter, I don't believe GPU-version is published. You have to build it or just use docker.

I did package installation from PyPI. Thanks for bringing this to my attention. @sskorol Will try with Docker. I would also like to try to build the sources with GPU support out-of-the-box. May I ask where can I obtain such source files?

**Comment by sskorol** [comment]
@scotgopal Vosk sources are here, in this repo. However, it's not a trivial process, as you have to build not only Vosk with GPU support, but also a Kaldi. It's better to take a look at available dockerfiles as they have all the required build steps that could be easily migrated to shell scripts for local installation. You can use [this one](https://github.com/sskorol/vosk-api-gpu/blob/main/pc/Dockerfile) as a reference.

**Comment by awesomer** [comment]
I apologize in advance if it's more appropriate for me to have filed this as a separate issue, but I have what seems like a closely related goal, that of running vosk-transcriber with GPU support under WSL2 with CUDA working properly. I am commenting here in the hopes of people with the same goal finding this issue via google as I did.

My situation is as follows :

a) I know that CUDA is working properly in my WSL2 environment because I've used it with demucs and checked its status recently with nvidia-smi.
b) I have compiled and installed both kaldi with --use-cuda and vosk-api with HAVE_CUDA=1, by manually running the commands contained within the docker file at https://github.com/sskorol/vosk-api-gpu/blob/main/pc/Dockerfile
c) I have tried running "from vosk import GpuInit, BatchModel" / "GpuInit()" / "model = BatchModel()" in an interactive python shell, and received the same "Exception: Failed to create a model" as above.
d) I have also tried running vosk-transcriber, and do not see any GPU related debug output or vosk-related GPU usage on my system.

Am I missing something, or is my error in believing that vosk-transcriber will use CUDA with this setup?

If this is my error, I am relatively content to use the docker + server deploy, but I feel very close to not needing the docker + server dependencies and would prefer to run without them. Thanks again for this amazing software.

Edit : I grepped for GpuInit in transcriber.py, and upon not finding it, tried to kludge it in by :

a) modifying the vosk import on line 13 to read : "from vosk import KaldiRecognizer, Model, GpuInit"
b) calling GpuInit() within pool_worker on line 142, right before "rec = KaldiRecognizer(self.model, SAMPLE_RATE)"

With no apparent change/benefit.
